### PR TITLE
[SPARK-43665][CONNECT][PS] Enable PandasSQLStringFormatter.vformat to work with Spark Connect

### PR DESCRIPTION
**Pull request by itholic**
### What changes were proposed in this pull request?

This PR aims enabling SQL parity test `test_sql_with_python_objects` for pandas API on Spark with Spark Connect.

### Why are the changes needed?

To increase the API coverage for pandas API on Spark with Spark Connect.

### Does this PR introduce _any_ user-facing change?

This enables `ps.sql` with Python objects.

### How was this patch tested?

Reuse the existing SQL tests.

### CODE REVIEW

**Comment by itholic** [comment]
cc @zhengruifeng FYI

File: python/pyspark/pandas/sql_formatter.py [hunk 1]
```diff
@@ -265,7 +266,10 @@ def _convert_value(self, val: Any, name: str) -> Optional[str]:
             val._to_spark().createOrReplaceTempView(df_name)
             return df_name
         elif isinstance(val, str):
-            return lit(val)._jc.expr().sql()  # for escaped characters.
+            if is_remote():
+                return f"'{val}'"
```
Comment (zhengruifeng) [inline python/pyspark/pandas/sql_formatter.py:270]: I think there is behavior change:

```
In [1]: from pyspark.sql.functions import *

In [2]: val = "a'''c''d"

In [3]: lit(val)._jc.expr().sql()
Out[3]: "'a\\'\\'\\'c\\'\\'d'"
```

Comment (itholic) [reply python/pyspark/pandas/sql_formatter.py]: Added test case with updating the code to handle escaped characters more properly. Thanks!

File: python/pyspark/pandas/sql_formatter.py [hunk 2]
```diff
@@ -265,7 +266,14 @@ def _convert_value(self, val: Any, name: str) -> Optional[str]:
             val._to_spark().createOrReplaceTempView(df_name)
             return df_name
         elif isinstance(val, str):
-            return lit(val)._jc.expr().sql()  # for escaped characters.
+            if is_remote():
+                special_characters = ["\\", "'", '"', "\0", "\b", "\n", "\r", "\t"]
+                escaped_val = val
+                for char in special_characters:
+                    escaped_val = escaped_val.replace(char, "\\" + char)
+                return f"'{escaped_val}'"
```
Comment (HyukjinKwon) [inline python/pyspark/pandas/sql_formatter.py:274]: This is too hacky. Is this exactly matched with JVM implementaton?

Comment (zhengruifeng) [reply python/pyspark/pandas/sql_formatter.py:274]: I cannot find a workaround based on DF apis, not sure whether a new proto message is needed

Comment (itholic) [reply python/pyspark/pandas/sql_formatter.py:274]: Yeah, actually I just refer the implementation from JVM to handle the `str` as below:

```scala
    case (v: UTF8String, StringType) =>
      // Escapes all backslashes and single quotes.
      "'" + v.toString.replace("\\", "\\\\").replace("'", "\\'") + "'"
```

Seems like they only care about the single quotation mark, but I additionally add some more possible special characters into `special_characters`.

Comment (HyukjinKwon) [reply python/pyspark/pandas/sql_formatter.py:274]: Okay, can you leave a comment here that this implementation matches with JVM side?

**Review by HyukjinKwon — APPROVED** [review]

File: python/pyspark/pandas/sql_formatter.py [hunk 3]
```diff
@@ -265,7 +266,14 @@ def _convert_value(self, val: Any, name: str) -> Optional[str]:
             val._to_spark().createOrReplaceTempView(df_name)
             return df_name
         elif isinstance(val, str):
-            return lit(val)._jc.expr().sql()  # for escaped characters.
+            if is_remote():
+                special_characters = ["\\", "'", '"', "\0", "\b", "\n", "\r", "\t"]
```
Comment (zhengruifeng) [inline python/pyspark/pandas/sql_formatter.py:270]: let's exactly match `"'" + v.toString.replace("\\", "\\\\").replace("'", "\\'") + "'"` in the JVM side

Comment (itholic) [reply python/pyspark/pandas/sql_formatter.py:270]: Sure. let me update the comments. Thanks!

**Review by zhengruifeng — APPROVED** [review]

File: python/pyspark/pandas/sql_formatter.py [hunk 4]
```diff
@@ -265,7 +266,13 @@ def _convert_value(self, val: Any, name: str) -> Optional[str]:
...
             return df_name
         elif isinstance(val, str):
-            return lit(val)._jc.expr().sql()  # for escaped characters.
+            if is_remote():
+                # This is matched to behavior from JVM implementation.
+                # See `sql` definition from `sql/catalyst/src/main/scala/org/apache/spark/
+                # sql/catalyst/expressions/literals.scala`
+                return "'" + val.replace("\\", "\\\\").replace("'", "\\'") + "'"
+            else:
+                return lit(val)._jc.expr().sql()  # for escaped characters.
```
Comment (HyukjinKwon) [inline python/pyspark/pandas/sql_formatter.py:275]:
```suggestion
            # This is matched to behavior from JVM implementation.
            # See `sql` definition from `sql/catalyst/src/main/scala/org/apache/spark/
            # sql/catalyst/expressions/literals.scala`
            return "'" + val.replace("\\", "\\\\").replace("'", "\\'") + "'"
```

Comment (HyukjinKwon) [reply python/pyspark/pandas/sql_formatter.py:275]: Actually let's remove the jvm access here then.

Comment (itholic) [reply python/pyspark/pandas/sql_formatter.py]: Sounds good

**Review by HyukjinKwon — APPROVED** [review]

**Comment by itholic** [comment]
Seems like the CI broken. The failure is not related to my changes.

**Comment by zhengruifeng** [comment]
merged to master

**Pull request closed without merging**